native/src/main/java/io/ballerina/stdlib/data/jsondata/FromString.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2024, WSO2 LLC. (https://www.wso2.com).
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.ballerina.stdlib.data.jsondata;
+
+import io.ballerina.runtime.api.PredefinedTypes;
+import io.ballerina.runtime.api.TypeTags;
+import io.ballerina.runtime.api.creators.TypeCreator;
+import io.ballerina.runtime.api.creators.ValueCreator;
+import io.ballerina.runtime.api.types.ReferenceType;
+import io.ballerina.runtime.api.types.Type;
+import io.ballerina.runtime.api.types.UnionType;
+import io.ballerina.runtime.api.utils.TypeUtils;
+import io.ballerina.runtime.api.values.BDecimal;
+import io.ballerina.runtime.api.values.BError;
+import io.ballerina.runtime.api.values.BString;
+import io.ballerina.runtime.api.values.BTypedesc;
+import io.ballerina.stdlib.data.jsondata.utils.DiagnosticErrorCode;
+import io.ballerina.stdlib.data.jsondata.utils.DiagnosticLog;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * Native implementation of data:fromStringWithType(string).
+ *
+ * @since 0.1.0
+ */
+public class FromString {
+
+    private static final HashMap<Integer, Integer> TYPE_PRIORITY_ORDER = new HashMap<>() {{
+        int precedence = 0;
+        put(TypeTags.INT_TAG, precedence++);
+        put(TypeTags.FLOAT_TAG, precedence++);
+        put(TypeTags.DECIMAL_TAG, precedence++);
+        put(TypeTags.NULL_TAG, precedence++);
+        put(TypeTags.BOOLEAN_TAG, precedence++);
+        put(TypeTags.JSON_TAG, precedence++);
+        put(TypeTags.STRING_TAG, precedence);
+    }};
```

Comment:
I am not sure if it is worth to use initialization block just to increase the precedence. 
```suggestion
    private static final Map<Integer, Integer> TYPE_PRIORITY_ORDER = Map.ofEntries(
        Map.entry(TypeTags.INT_TAG, 0),
        Map.entry(TypeTags.FLOAT_TAG, 1),
        Map.entry(TypeTags.DECIMAL_TAG, 2),
        Map.entry(TypeTags.NULL_TAG, 3),
        Map.entry(TypeTags.BOOLEAN_TAG, 4),
        Map.entry(TypeTags.JSON_TAG, 5),
        Map.entry(TypeTags.STRING_TAG, 6)
);
```

Reply:
Hard coding the precedence is not best practice. Right? assume if we are going to introduce new data type and rearrange the precedence, in such cases incrementing reduce human error. WDYT?

Reply:
I agree with your point. However, there are two engineering perspectives that you need to consider.

1. How frequently will you be changing the order? JSON tags are standardized and do not have the potential to change, so I am not sure if we need to worry about this.
2. Is the performance trade-off worth it (both performance and storage)? 
Compared to mine, your initialization does not address immutability and compact size. Although immutability has a minor impact on single-threaded applications, it still provides the potential for the compiler to provide aggressive optimizations. Also, if you are to use HashMap, it's better if we can limit the size since we already know the max count, `new HashMap(TAGS_COUNT) {{`.

Also, why are we opting to use a map here? Since the size of the Map is guaranteed to be small, there is no major performance advantage of using a HashMap here. You can also have the items in an array (so you do not have to manually define the precedence) and use the `indexOf` method to obtain the priority. (source from Gemini for the performance difference between the two approaches: https://g.co/gemini/share/31fed1c8c4be)